internal/services/v1/watch_test.go

```diff
@@ -0,0 +1,206 @@
+				ObjectType: resourceObjType,
+				ObjectId:   resourceObjID,
+			},
+			Relation: relation,
+			Subject: &v1.SubjectReference{
+				Object: &v1.ObjectReference{
+					ObjectType: subObjType,
+					ObjectId:   subObjectID,
+				},
+			},
+		},
+	}
+}
+
+func TestWatch(t *testing.T) {
+	testCases := []struct {
+		name              string
+		objectTypesFilter []string
+		startCursor       *v1.ZedToken
+		mutations         []*v1.RelationshipUpdate
+		expectedCode      codes.Code
+		expectedUpdates   []*v1.RelationshipUpdate
+	}{
+		{
+			name:         "unfiltered watch",
+			expectedCode: codes.OK,
+			mutations: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "folder", "folder1", "viewer", "user", "user1"),
+			},
+			expectedUpdates: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "folder", "folder1", "viewer", "user", "user1"),
+			},
+		},
+		{
+			name:              "watch with objectType filter",
+			expectedCode:      codes.OK,
+			objectTypesFilter: []string{"document"},
+			mutations: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "folder", "folder1", "viewer", "user", "user1"),
+			},
+			expectedUpdates: []*v1.RelationshipUpdate{
+				update(v1.RelationshipUpdate_OPERATION_CREATE, "document", "document1", "viewer", "user", "user1"),
+			},
+		},
+		{
+			name:         "invalid zedtoken",
+			startCursor:  &v1.ZedToken{Token: "bad-token"},
+			expectedCode: codes.InvalidArgument,
+		},
+		{
+			name:         "empty zedtoken fails validation",
+			startCursor:  &v1.ZedToken{Token: ""},
+			expectedCode: codes.InvalidArgument,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			require := require.New(t)
+
+			rawDS, err := memdb.NewMemdbDatastore(0, 0, memdb.DisableGC, 0)
+			require.NoError(err)
+
+			ds, _ := testfixtures.StandardDatastoreWithSchema(rawDS, require)
+
+			revision, err := ds.SyncRevision(context.Background())
+			require.NoError(err)
+			require.True(revision.GreaterThan(decimal.Zero))
+
+			client, stop := newWatchServicer(require, ds)
+			defer stop()
+
+			cursor := zedtoken.NewFromRevision(revision)
+			if tc.startCursor != nil {
+				cursor = tc.startCursor
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			stream, err := client.Watch(ctx, &v1.WatchRequest{
+				OptionalObjectTypes: tc.objectTypesFilter,
+				OptionalStartCursor: cursor,
+			})
+			require.NoError(err)
+
+			if tc.expectedCode == codes.OK {
+
+				updatesChan := make(chan []*v1.RelationshipUpdate, len(tc.mutations))
+
+				go func() {
+					defer func() {
```

Comment:
`defer close(updatesChan)`?

Reply:
Yeah, is it wrong to close the updatesChan after we're done writing to it?

Reply:
Ah, sorry you're saying streamlining the defer. 👍

Reply:
Probably not, no

Reply:
Fixed.